packages/block-editor/src/store/reducer.js

```diff
@@ -645,17 +684,56 @@ export const blocks = flow(
 			case 'REMOVE_BLOCKS':
 				return flow( [
 					// Remove inner block ordering for removed blocks
-					( nextState ) => omit( nextState, action.clientIds ),
+					( nextState ) => omit( nextState, action.removedClientIds ),
 
 					// Remove deleted blocks from other blocks' orderings
 					( nextState ) => mapValues( nextState, ( subState ) => (
-						without( subState, ...action.clientIds )
+						without( subState, ...action.removedClientIds )
 					) ),
 				] )( state );
 		}
 
 		return state;
 	},
+
+	// While technically redundant data as the inverse of `order`, it serves as
+	// an optimization for the selectors which derive the ancestry of a block.
+	parents( state = {}, action ) {
+		switch ( action.type ) {
+			case 'RESET_BLOCKS':
+				return mapBlockParents( action.blocks );
+
+			case 'RECEIVE_BLOCKS':
+				return {
+					...state,
+					...mapBlockParents( action.blocks ),
+				};
+
+			case 'INSERT_BLOCKS':
+				return {
+					...state,
+					...mapBlockParents( action.blocks, action.rootClientId || '' ),
+				};
+
+			case 'MOVE_BLOCK_TO_POSITION': {
+				return {
+					...state,
+					[ action.clientId ]: action.toRootClientId || '',
+				};
+			}
+
+			case 'REPLACE_BLOCKS':
+				return {
+					...omit( state, action.replacedClientIds ),
+					...mapBlockParents( action.blocks, state[ action.clientIds[ 0 ] ] ),
```

Comment:
Re: My previous comment, I guess we still reference the original `clientIds` here. Can you clarify what you're doing here in referencing only the first of the original of the `clientIds` ?

Reply:
Basically we need the parent of the replaced blocks to assign it as parent of the inserted blocks. `replacedClientIds` contains more than just the "root" level which in theory could result in the wrong value here.